File: typescript/src/builder/director.ts
import type { Builder } from "./builder.ts";
import type { Product } from "./product.ts";

/** Owns the sequence of construction steps; knows nothing of the representation. */
export class Director {
  construct(builder: Builder): Product {
    builder.buildPartA();
    builder.buildPartB();
    return builder.getResult();
  }
}

File: python/src/patterns/builder/director.py
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .builder import Builder
    from .product import Product


class Director:
    """Owns the sequence of construction steps; knows nothing of the representation."""

    def construct(self, builder: Builder) -> Product:
        builder.build_part_a()
        builder.build_part_b()
        return builder.get_result()
